s3/authentication.go

```diff
@@ -0,0 +1,257 @@
+		CreateBucket            bool
+		BucketExists            bool
+		DeleteBucket            bool
+		GetObject               bool
+		HeadObject              bool
+		DeleteObject            bool
+		PutObject               bool
+		DeleteMulti             bool
+		CopyObject              bool
+		CreateMultipartUpload   bool
+		UploadPart              bool
+		ListMultipartUpload     bool
+		ListParts               bool
+		AbortMultipartUpload    bool
+		CompleteMultipartUpload bool
+	}
+
+	contextKey int
+)
+
+var (
+	permissionKey contextKey
+
+	// rootPerms are used for requests that were successfully authenticated
+	// using v4 signatures.
+	rootPerms = permissions{
+		ListBuckets:             true,
+		ListBucket:              true,
+		CreateBucket:            true,
+		BucketExists:            true,
+		DeleteBucket:            true,
+		GetObject:               true,
+		HeadObject:              true,
+		DeleteObject:            true,
+		PutObject:               true,
+		DeleteMulti:             true,
+		CopyObject:              true,
+		CreateMultipartUpload:   true,
+		UploadPart:              true,
+		ListMultipartUpload:     true,
+		ListParts:               true,
+		AbortMultipartUpload:    true,
+		CompleteMultipartUpload: true,
+	}
+
+	// noAccessPerms grant access to nothing.
+	noAccessPerms = permissions{}
+)
+
+func newAuthenticatedBackend(b *s3, keyPairs map[string]string) *authenticatedBackend {
+	signature.StoreKeys(keyPairs)
+	return &authenticatedBackend{
+		backend:    b,
+		v4AuthPair: keyPairs,
+	}
+}
+
+func (b *authenticatedBackend) applyBucketPolicy(ctx context.Context, bucketName string, p *permissions) error {
+	bucket, err := b.backend.b.Bucket(ctx, bucketName)
+	if err != nil && strings.Contains(err.Error(), api.ErrBucketNotFound.Error()) {
+		return gofakes3.BucketNotFound(bucketName)
+	} else if err != nil {
+		return gofakes3.ErrorMessage(gofakes3.ErrInternal, err.Error())
+	}
+	if bucket.Policy.PublicReadAccess {
+		p.ListBucket = true
+		p.BucketExists = true
+		p.GetObject = true
+		p.HeadObject = true
+	}
+	return nil
+}
+
+func (b *authenticatedBackend) permsFromCtx(ctx context.Context, bucket string) permissions {
+	perms := noAccessPerms
+	if p, ok := ctx.Value(permissionKey).(*permissions); ok {
+		perms = *p
+	}
+	if bucket != "" {
+		b.applyBucketPolicy(ctx, bucket, &perms)
+	}
+	return perms
+}
+
+func (b *authenticatedBackend) AuthenticationMiddleware(handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, rq *http.Request) {
+		perms := noAccessPerms
+		if len(b.v4AuthPair) > 0 {
+			if rq.Header.Get("Authorization") == "" {
+				// No auth header, we use guest permissions.
+				// TODO: guest permissions
+			} else if result := signature.V4SignVerify(rq); result != signature.ErrNone {
+				// Authentication attempted but failed.
+				resp := signature.GetAPIError(result)
+				w.WriteHeader(resp.HTTPStatusCode)
+				w.Header().Add("content-type", "application/xml")
```

Comment:
```suggestion
				w.Header().Add("Content-Type", "application/xml")
```
probably doesn't matter but can't hurt